aten/src/ATen/native/UnaryOps.cpp

```diff
@@ -611,9 +611,9 @@ Tensor& arctanh_out(const Tensor& self, Tensor& result) { return at::atanh_out(r
 Tensor arctanh(const Tensor& self) { return self.atanh(); }
 Tensor& arctanh_(Tensor& self) { return self.atanh_(); }
 
-Tensor& square_out(const Tensor& self, Tensor& result) { return at::pow_out(result, self, 2); }
-Tensor square(const Tensor& self) { return at::pow(self, 2); }
-Tensor& square_(Tensor& self) { return self.pow_(2); }
+Tensor& square_out(const Tensor& self, Tensor& result) { return at::mul_out(result, self, self); }
+Tensor square(const Tensor& self) { return at::mul(self, self); }
+Tensor& square_(Tensor& self) { return self.mul_(self); }
```

Comment:
We probably want to do `temp = self.clone(); self.copy_(temp * temp)`

The issue is that when you do `a.mul_(b)` autograd needs to save `b` so it can compute gradient wrt a during backward and variables that are saved for backward cannot be modified in-place.

Reply:
But if we do `temp = self.clone(); self.copy_(temp * temp)` that will significantly slow down in-place `square_` when autograd isn't needed?

Reply:
That's a good point - we'd need to only do that if we know we'll need autograd.

Also we only want to do this if we know that the overhead we're adding is less than what we're saving by changing pow to mul. Otherwise, sticking with pow_ would be better.
 - out-of-place clone + out-of-place mul + in-place copy (
 - out-of-place clone + in-place pow

To actually fully take advantage of mul being faster for the in-place case, we'd need a good amount of autograd codegen changes (to changing what we save/use in formulas depending on whether other is the same as self) or a manual variable type kernel, which we generally want to avoid.

Reply:
I understand now that the interaction with AD makes this more complex. Maybe there's still a way to get some of the speed by improving `pow`.